### ISSUE TITLE
Rails 6 will remove success? predicate in favor of successful?

### ISSUE THREAD

**Issue by ybart** [rspec/rspec-rails#1857]
While running against Rails 5.2.0.alpha, I encountered the following message when running my test suite:

```
.DEPRECATION WARNING: The success? predicate is deprecated and will be removed in Rails 6.0. Please use successful? as provided by Rack::Response::Helpers. (called from matches? at rspec/rails/matchers/have_http_status.rb:263)
```

The message is coming from https://github.com/rails/rails/pull/30104 (https://github.com/rails/rails/commit/af3500b18817a48d7ec83812a0d4869f3fe33b2f)

**Comment by trev** [comment]
Hi @ybart, happy to help with this. What's your question?

**Comment by ybart** [comment]
I'm not sure how to deal with this. Maybe we may edit https://github.com/rspec/rspec-rails/blob/f33dde7519433f59fa21778aba953f2237dfacdb/lib/rspec/rails/matchers/have_http_status.rb#L245 to allow statuses referenced here http://www.rubydoc.info/gems/rack/Rack/Response/Helpers as well.

**Comment by trev** [comment]
Basically the whole [`GenericStatus`](https://github.com/rspec/rspec-rails/blob/f33dde7519433f59fa21778aba953f2237dfacdb/lib/rspec/rails/matchers/have_http_status.rb#L20) class functionality needs to be deprecated and removed in lockstep with Rails 6.0.

**Comment by trev** [comment]
Probably something like:

```ruby
        def self.matcher_for_status(target)
          if GenericStatus.valid_statuses.include?(target)
            RSpec.deprecate(
              GenericStatus.valid_statuses.map { |c| "`:#{c}`" }.join(', '),
              replacement: "the Rack::Response::Helpers predicates as symbols",
            )
            GenericStatus.new(target)
          elsif Symbol === target
            SymbolicStatus.new(target)
          else
            NumericCode.new(target)
          end
        end
```

alternatively we could just leave as is since the Rails deprecation takes care of it.

In your case you're probably using something like `expect(response).to have_http_status(:success)` in your tests. To get rid of the deprecation just use `expect(response).to have_http_status(:successful)`

**Comment by ybart** [comment]
Thanks for the answer, just tried that and got the following error :

``` 
1) HomeController GET #show returns http success
     Failure/Error: expect(response).to have_http_status(:successful)

     ArgumentError:
       Invalid HTTP status: :successful
     # ./spec/controllers/home_controller_spec.rb:11:in `block (3 levels) in <top (required)>'
```

I thinks this is because `SymbolicStatus` matches against `Rack::Utils::SYMBOL_TO_STATUS_CODE`, which does not contain :successful.

**Comment by trev** [comment]
Good point. In the interim you could do `expect(response).to have_http_status(200)` or whatever successful response code you'd expect.

Perhaps a more adequate solution would be to make these [Rack Helpers](https://github.com/rack/rack/blob/master/lib/rack/response.rb#L111-L117) the next GenericStatus.valid_statuses whilst deprecating the old ones.

**Comment by ybart** [comment]
I'm writing a PR for the more adequate solution.

However, I am not sure about `redirect`: [This commit](https://github.com/rails/rails/commit/4d4d764ab688af4a0d213ab2d0dc7f9b811ab451#diff-c5366c6770ca1168af3972fcbcff7df6) removed it in favor of Rack redirect one, but semantics are not exactly the same, redirection means [300..399] and redirect means [301, 302, 307, 308].

**Comment by jonekdahl** [comment]
Is this fixed by #1951?

**Comment by JonRowe** [comment]
Closed by #1951
